File: Java/Guias/guia3/ej8_CellPhone/FriendCellPhoneBill.java
package Guias.guia3.ej8_CellPhone;

public class FriendCellPhoneBill extends CellPhoneBill {

    private String[] friends;
    private int friendsId;
    private int friendsLimit;
    private double cost;
    private static final double defaultFriendCostPerSecond = 0;

    FriendCellPhoneBill(String number, int limit) {
        super(number);
        this.friendsLimit = limit;
        this.friendsId = 0;
        this.friends = new String[friendsLimit];
        this.cost = defaultFriendCostPerSecond;
    }

    public void setCost(double cost) {
        this.cost = cost;
    }

    public void addFriend(String friend) {
        if (friendsId < friendsLimit && !containsFriend(friend)) {
            this.friends[friendsId++] = friend;
        } else {
            System.out.println("Error, ya llego al limite de amigos agregados");
        }
    }

    private boolean containsFriend(String number) {
        for (int i = 0; i < friendsId; i++) {
            if (friends[i].compareTo(number) == 0) {
                return true;
            }
        }
        return false;
    }

    public void removeFriend(String number) {
        boolean found = false;
        for (int i = 0; i < friendsId && !found; i++) {
            if (friends[i].compareTo(number) == 0) {
                friends[i] = null;
                friendsId--;
                found = true;
            }
        }
    }

    @Override
    public double processBill() {
        double total = 0;
        for (int i = 0; i < callsIndex; i++) {
            boolean friendlyCall = false;
            for (int j = 0; j < friendsId && !friendlyCall; j++) {
                if (calls[i].getTo().compareTo(friends[j]) == 0) {
                    friendlyCall = true;
                }
            }
            total += friendlyCall ? calls[i].getCost() * cost : calls[i].getCost();
        }
        return total;
    }

}

File: RUBY/Guias/Guia10/Ej4_FriendCellPhone/friend_cell_phone_bill.rb
# frozen_string_literal: true
require_relative 'cell_phone_bill'
class FriendCellPhoneBill < CellPhoneBill

  def initialize(number, limit)
    super(number)
    @limit = limit
    @friends = []
    @@cost_friends_per_second = 0
  end

  def self.change_cost(cost)
    @@cost_friends_per_second = cost
  end

  def add_friend(number)
    raise 'Friend already exists' if @friends.include?(number)
    raise 'Too many friends' if @friends.length >= @limit
    @friends.push(number)
  end

  def remove_friend(number)
    raise 'Friend does not exist' unless @friends.include?(number)
    @friends.delete(number)
  end

  def process_bill
    @calls.map{ |c| @friends.include?(c.to) ? c.cost * @@cost_friends_per_second : c.cost }.reduce(0, :+)
  end
end
